File: leetcode/530.minimum-absolute-difference-in-bst/ts/getminimumdifference.ts
class TreeNode {
    val: number
    left: TreeNode | null
    right: TreeNode | null
    constructor(val?: number, left?: TreeNode | null, right?: TreeNode | null) {
        this.val = (val === undefined ? 0 : val)
        this.left = (left === undefined ? null : left)
        this.right = (right === undefined ? null : right)
    }
}

const getMinimumDifference = (root: TreeNode | null): number => {
    let prev: TreeNode | null = null
    let minDiff = Infinity

    const traversal = (node: TreeNode | null): void => {
        if (node === null) return

        traversal(node.left)

        if (prev !== null) {
            const diff = Math.abs(node.val - prev.val)
            minDiff = Math.min(minDiff, diff)
        }

        prev = node

        traversal(node.right)
    }

    traversal(root)
    return minDiff
}

export default getMinimumDifference

File: leetcode/530.minimum-absolute-difference-in-bst/go/getminimumdifference.go
package getmindiff

import (
	"math"
)

type TreeNode struct {
	Val   int
	Left  *TreeNode
	Right *TreeNode
}

func getMinimumDifference(root *TreeNode) int {
	var prev *TreeNode
	minDiff := math.MaxInt

	var traversal func(node *TreeNode)

	traversal = func(node *TreeNode) {
		if node == nil {
			return
		}

		traversal(node.Left)

		if prev != nil {
			diff := abs(node.Val - prev.Val)
			if diff < minDiff {
				minDiff = diff
			}
		}

		prev = node

		traversal(node.Right)
	}

	traversal(root)
	return minDiff
}

func abs(v int) int {
	if v >= 0 {
		return v
	}
	return -v
}
